lib/msf/core/post/file.rb

```diff
@@ -376,33 +389,104 @@ def write_file(file_name, data)
       fd.close
     elsif session.respond_to? :shell_command_token
       if session.platform == 'windows'
-        session.shell_command_token("echo #{data} > \"#{file_name}\"")
+        if is_safe?(data)
+          return win_ansi_write_file(file_name, data)
+        else
+          return win_bin_write_file(file_name, data)
+        end
       else
-        _write_file_unix_shell(file_name, data)
+        return _write_file_unix_shell(file_name, data)
       end
     end
-    true
   end
 
+  # Windows ANSI file write for shell sessions. Writes given object content to a remote file.
   #
-  # Platform-agnostic file append. Appends given object content to a remote file.
-  # Returns Boolean true if successful
+  # NOTE: *This is not binary-safe on Windows shell sessions!*
+  #
+  # @param file_name [String] Remote file name to write
+  # @param data [String] Contents to put in the file
+  # @param max_chunk [int] max size for the data chunk to write at a time
+  # @return [void]
+  def win_ansi_write_file(file_name, data, max_chunk = 5000)
+    start_index = 0
+    write_length =[max_chunk, data.length].min
+    session.shell_command_token("echo | set /p=\"#{data[0, write_length]}\"> \"#{file_name}\"")
```

Comment:
The ascii binary might have `"` quotes in it, will this cause an issue? 👀

Reply:
I added the quotation to the bad char check.